components/director/pkg/applicationtenancy/directive.go

```diff
@@ -0,0 +1,234 @@
+}
+
+// ApplicationService is responsible for the service-layer application operations.
+//go:generate mockery --name=ApplicationService --output=automock --outpkg=automock --case=underscore --disable-version-string
+type ApplicationService interface {
+	ListAll(ctx context.Context) ([]*model.Application, error)
+}
+
+type directive struct {
+	transact      persistence.Transactioner
+	tenantService BusinessTenantMappingService
+	appService    ApplicationService
+}
+
+// NewDirective creates a directive object
+func NewDirective(transact persistence.Transactioner, tenantService BusinessTenantMappingService, appService ApplicationService) *directive {
+	return &directive{
+		transact:      transact,
+		tenantService: tenantService,
+		appService:    appService,
+	}
+}
+
+// SynchronizeApplicationTenancy handles graphql.EventTypeNewApplication, graphql.EventTypeNewSingleTenant, and graphql.EventTypeNewMultipleTenants events.
+// In EventTypeNewApplication we extract the customer parent of the owner of the new application. We give access to accounts under that customer to access the application.
+// In EventTypeNewSingleTenant we get the new tenant's customer parent. If the new tenant is account we give him access to all Atom applications under the customer.
+// In EventTypeNewMultipleTenants we do the same as for EventTypeNewSingleTenant event but for multiple new tenants.
+func (d *directive) SynchronizeApplicationTenancy(ctx context.Context, _ interface{}, next gqlgen.Resolver, eventType graphql.EventType) (res interface{}, err error) {
+	resp, err := next(ctx)
+	if err != nil {
+		log.C(ctx).WithError(err).Errorf("An error occurred while processing request: %s", err.Error())
+		return resp, err
+	}
+
+	tx, err := d.transact.Begin()
+	if err != nil {
+		log.C(ctx).WithError(err).Errorf("An error occurred while opening database transaction")
+		return nil, apperrors.NewInternalError("Unable to initialize database transaction: %s", err.Error())
+	}
+	defer d.transact.RollbackUnlessCommitted(ctx, tx)
+
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	log.C(ctx).Debugf("Preparing tenant access creation for event: %s", eventType)
+
+	if eventType == graphql.EventTypeNewApplication {
+		tntIDFromContext, err := tenant.LoadFromContext(ctx)
+		if err != nil {
+			log.C(ctx).WithError(err).Errorf("An error occurred while loading tenant from context")
+			return nil, err
+		}
+
+		tntModel, err := d.tenantService.GetTenantByID(ctx, tntIDFromContext)
+		if err != nil {
+			log.C(ctx).WithError(err).Errorf("An error occurred while getting tenant with ID %s", tntIDFromContext)
+			return nil, errors.Wrapf(err, "while fetching tenant with id: %s", tntIDFromContext)
+		}
+
+		log.C(ctx).Debugf("Found a matching tenant in the database: %s", tntModel.ID)
+
+		if !(tntModel.Type == tenantpkg.ResourceGroup || tntModel.Type == tenantpkg.Folder || tntModel.Type == tenantpkg.Organization) {
```

Comment:
tntModel.Type != tenantpkg.ResourceGroup && tntModel.Type != tenantpkg.Folder && tntModel.Type != tenantpkg.Organization) is more readable for me